orttraining/orttraining/core/optimizer/gist_encode_decode.cc

```diff
@@ -45,74 +45,162 @@ static std::vector<GraphEdgeHelper> GetNodeOutputEdges(const Node& node) {
+      }
+    }
+  }
 
-  std::string encode_node_name = graph.GenerateNodeName(GIST_ENCODER_NODE_NAME_BASE);
-  for (int i = 0; i < curr_node_output_arg->Shape()->dim_size(); i++) {
-    bool_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(curr_node_output_arg->Shape()->dim(i).dim_value());
+  if (decode_map.empty()) {
+    return false;
   }
-  auto& encode_output_def_compressed_arg = graph.GetOrCreateNodeArg(encode_node_name, &bool_tensor);
-  auto& encode_output_def_uncompressed_arg = graph.GetOrCreateNodeArg(encode_node_name + "_identity", curr_node_output_arg->TypeAsProto());
-  auto& encode = graph.AddNode(encode_node_name, compression_type + "Encoder", "Encode", {curr_node_output_arg}, {&encode_output_def_uncompressed_arg, &encode_output_def_compressed_arg}, {}, kMSDomain);
 
-  std::string decode_arg_name = graph.GenerateNodeName(GIST_DECODER_NODE_NAME_BASE);
-  auto& decode_output_def_uncompressed_arg = graph.GetOrCreateNodeArg(decode_arg_name, curr_node_output_arg->TypeAsProto());
-  auto& decode_output_def_dummy_arg = graph.GetOrCreateNodeArg(decode_arg_name + "_late_dec", curr_node_output_arg->TypeAsProto());
-  auto& decode = graph.AddNode(decode_arg_name, compression_type + "Decoder", "Decode", {&decode_output_def_dummy_arg, &encode_output_def_compressed_arg}, {&decode_output_def_uncompressed_arg}, {}, kMSDomain);
+  std::string user_compression_type = compression_type;
 
-  bool early_encoding = false;
-  bool late_decoding = false;
-  for (auto& output_edge : output_edges) {
-    Node* node_dst = graph.GetNode(output_edge.dst_node);
-    if (node_dst->Description() == "Backward pass" && (node_dst->OpType() == "ReluGrad")) {
-      graph.AddEdge(output_edge.src_node, encode.Index(), output_edge.src_arg_index, 0);
-      graph.AddEdge(encode.Index(), decode.Index(), 1, 1);
-      graph.AddEdge(decode.Index(), output_edge.dst_node, 0, output_edge.dst_arg_index);
-      std::vector<GraphEdgeHelper> input_edges_dst = GetNodeInputEdges(*node_dst);
-      size_t i = 0;
-      while (!late_decoding && i < input_edges_dst.size()) {
-        if (graph.GetNode(input_edges_dst[i].src_node)->OpType() != curr_node.OpType()) {
-          graph.AddEdge(input_edges_dst[i].src_node, decode.Index(), input_edges_dst[i].src_arg_index, 0);
-          late_decoding = true;
+  // Each element in map corresponds to a stash activation
+  for (auto& st_act : decode_map) {
+    // Create compressed tensor
+    NodeArg* curr_node_output_arg = curr_node.MutableOutputDefs()[st_act.first];
+    ONNX_NAMESPACE::TypeProto compressed_tensor;
+    compression_type = user_compression_type;
+
+    // Override compression_type for lossless compression case(s) (eg. bool -> Pack1)
+    ONNX_NAMESPACE::DataType type_string = curr_node_output_arg->Type();
+    if (*type_string == "bool" || *type_string == "tensor(bool)") {
+      std::cout << "(Lossless) override compression type to Pack1 for tensor: " << curr_node_output_arg->Name() << std::endl;
```

Comment:
Use the `Logger` here (probably at level `INFO`) instead of `std::cout`.

Reply:
Updated to use logger.